pkg/pillar/cmd/nim/controllerdns_test.go

```diff
@@ -0,0 +1,212 @@
+					IP:   []byte{192, 168, 1, 1},
+					Mask: []byte{255, 255, 255, 0},
+				},
+				{
+					IP:   []byte{192, 168, 1, 2},
+					Mask: []byte{255, 255, 255, 0},
+				},
+			},
+			HwAddr: []byte{},
+			DNS: netmonitor.DNSInfo{
+				ResolvConfPath: "/etc/resolv.conf",
+				Domains:        []string{},
+				DNSServers: []net.IP{
+					{1, 0, 0, 1},
+					{8, 8, 8, 8},
+				},
+			},
+		},
+		{
+			Attrs: netmonitor.IfAttrs{
+				IfIndex: 2,
+				IfName:  "ExpensiveIf",
+			},
+			IPAddrs: []*net.IPNet{{
+				IP:   []byte{6, 6, 6, 6},
+				Mask: []byte{255, 255, 255, 0},
+			}},
+			HwAddr: []byte{},
+			DNS: netmonitor.DNSInfo{
+				ResolvConfPath: "/etc/resolv.conf",
+				Domains:        []string{},
+				DNSServers: []net.IP{
+					{0, 6, 6, 6},
+					{0, 7, 7, 7},
+				},
+			},
+		},
+	}
+
+	dpcManagerMock := dpcmanager.DevicePortConfigMock{}
+	deviceNetworkStatus := types.DeviceNetworkStatus{
+		CurrentIndex: 0,
+		Ports: []types.NetworkPortStatus{
+			{
+				IfName: mockInterface[0].Attrs.IfName,
+				AddrInfoList: []types.AddrInfo{
+					{Addr: mockInterface[0].IPAddrs[0].IP},
+					{Addr: mockInterface[0].IPAddrs[1].IP},
+				},
+			},
+			{
+				IfName: mockInterface[1].Attrs.IfName,
+				AddrInfoList: []types.AddrInfo{
+					{Addr: mockInterface[1].IPAddrs[0].IP},
+					{Addr: mockInterface[1].IPAddrs[1].IP},
+				},
+			},
+			{
+				IfName: mockInterface[2].Attrs.IfName,
+				Cost:   0,
+				AddrInfoList: []types.AddrInfo{
+					{Addr: net.IP{6, 6, 6, 6}},
+				},
+			},
+		},
+	}
+	dpcManagerMock.SetDNS(deviceNetworkStatus)
+
+	networkMonitor := netmonitor.MockNetworkMonitor{Log: log}
+	networkMonitor.AddOrUpdateInterface(mockInterface[0])
+	networkMonitor.AddOrUpdateInterface(mockInterface[1])
+	networkMonitor.AddOrUpdateInterface(mockInterface[2])
+	nim := &nim{
+		AgentBase:      agentbase.AgentBase{},
+		Log:            log,
+		Logger:         logger,
+		dpcManager:     &dpcManagerMock,
+		dpcReconciler:  nil,
+		networkMonitor: &networkMonitor,
+	}
+
+	return nim
+}
+
+func TestDnsResolve(t *testing.T) {
+	t.Parallel()
+
+	testHost := "255.255.255.255.nip.io"
+	expectedIP := net.IP{255, 255, 255, 255}
+	if testing.Short() {
+		t.Skipf("Skipping as connecting to the internet would take too much time and short tests are enabled")
+	}
+
+	nim := createTestNim()
+	res := nim.resolveWithSrcIP(testHost, net.IP{1, 1, 1, 1}, net.IP{0, 0, 0, 0})
+	if res == nil {
+		t.Skipf("could not resolve, skipping as probably the tests don't have internet connection of %s is down", testHost)
+	}
+	if !res[0].ip.Equal(expectedIP) {
+		t.Fatalf("resolving returned wrong IP address %+v, but should have been %+v", res, expectedIP)
+	}
+}
+
+func TestDnsResolveTimeout(t *testing.T) {
+	t.Parallel()
+
+	if testing.Short() {
+		t.Skipf("Skipping as timing out would take too much time and short tests are enabled")
+	}
+	nim := createTestNim()
+	res := nim.resolveWithSrcIP("www.google.com", net.IP{ /* example.com */ 93, 184, 216, 34}, net.IP{172, 17, 0, 3})
```

Comment:
I do not quite understand this test - does it use 93.184.216.34 (example.com) as DNS server and 172.17.0.3 as source IP? The Fatalf condition below talks about some local DNS server. Please make sure that unit tests are as independent from the host where they are executed as possible.

Reply:
The test tries to use example.com as DNS server and check that it is an invalid DNS server and therefore www.google.com cannot be resolved.
It is independent from the host, but expects example.com not being a DNS server.

Reply:
Also changed it from 172.17.0.3 to 0.0.0.0 in case it is not the first docker container running

Reply:
OK, now it is clear to me, thanks.